cdc/log.cc

```diff
@@ -519,6 +519,8 @@ api::timestamp_type find_timestamp(const schema& s, const mutation& m) {
                     [&] (collection_mutation_view_description mview) {
                 t = mview.tomb.timestamp;
                 if (t != api::missing_timestamp) {
+                    t += 1; // TODO: test this behavior
```

Comment:
Could you please explain what this is for and why it's needed?

Reply:
Yes please. Are you preserving the timestamp from the original collection mutation tombstone and trying to compensate for the adjustment there? If so, a more solid approach would be to use the timestamp or the rest of the collection mutation set, not adjusting by constant.

Reply:
There might be no cells in the collection mutation, just a tombstone.
This is the same adjustment that we're doing in the splitting logic:
```
            if (desc.tomb) {
                auto timestamp_and_ttl = std::pair(desc.tomb.timestamp + 1, gc_clock::duration(0));
                auto& updates = result[timestamp_and_ttl].nonatomic_updates;
                if (updates.empty() || updates.back().id != id) {
                    updates.push_back({id, {}});
                }
                updates.back().t = std::move(desc.tomb);
            }
```

Why are we doing this? I think it should be explained in detail in the PR cover letter (@jul-stas). For now let me explain it here.

Suppose we have a table `ks.t` with non-frozen collection column `X`.
The following statement:
```update ks.t using timestamp T set X = null where ...```
creates a tombstone for column `X` **with timestamp `T-1`**.
The following statement:
```update ks.t using timestamp T set X = X + {cell} where ...```
creates a tombstone for column `X` **with timestamp `T-1`** and a cell for column `X` with timestamp `T`.
Because we don't want to split the second update into two log rows, then for any timestamp `T`, **we put collection tombstones with timestamp `T` into the `T+1` bucket** (they will show up in the log with timestamp `T+1`). In particular, if we have both a tombstone and a bunch of cells, such that the timestamps of these cells are the tombstone's timestamp + 1, we put them together into a single log row (unless cells have TTLs, but that's a different story).
The user can then replay this log entry using an `update` statement with this timestamp.

Unfortunately, there is a small inconsistency. While:
```update ks.t using timestamp T set X = null where ...```
uses timestamp `T-1` for the tombstone,
```delete from ks.t using timestamp T where ...```
**uses timestamp `T`** :(
So if we're given a mutation with collection tombstone with timestamp `T`, what timestamp should we use for the `time` column? If we use `T`, it will be consistent with `delete`, but won't be consistent with `update`. If we use `T+1`, it will be consistent with `update`, but won't be consistent with `delete`. But **we have to choose something**. And the best thing to do is to use `T+1` to be consistent with `update`, because it causes less splitting.
We simply need to document the following for the user: when replaying the log, don't use `delete` statement, use `update` and then `cdc$time` provides you the correct timestamp. If you really want to use `delete` for some reason (you don't need to), you need to adjust the timestamp from `cdc$time` by +1.

Reply:
> Overwriting or inserting a collection using timestamp T is a process with
> following steps:
> 
>  inserting a row marker (if applicable) with timestamp T;
>  writing a collection tombstone with timestamp T-1;
>  writing the new collection value with timestamp T.

@jul-stas not true for `DELETE`, as explained above.